pkg/controllers/metrics/pod/controller.go

```diff
@@ -0,0 +1,146 @@
+
+	// 1. Has the provisioner been deleted?
+	provisioner := v1alpha4.Provisioner{}
+	if err := c.KubeClient.Get(ctx, req.NamespacedName, &provisioner); err != nil {
+		if !errors.IsNotFound(err) {
+			// Unable to determine existence of the provisioner, try again later.
+			return reconcile.Result{Requeue: true}, err
+		}
+
+		// The provisioner has been deleted. Reset all the associated counts to zero.
+		if err := publishPodCounts(provisioner.Name, map[string][]v1.Pod{}, []*v1.Pod{}); err != nil {
+			// One or more metrics were not zeroed. Try again later.
+			return reconcile.Result{Requeue: true}, err
+		}
+
+		// Since the provisioner is gone, do not requeue.
+		return reconcile.Result{}, nil
+	}
+
+	// 2. Update pod counts associated with the provisioner.
+	podsByZone, err := c.podsByZone(ctx, &provisioner)
+	if err != nil {
+		return reconcile.Result{Requeue: true}, err
+	}
+	provisionablePods, err := c.Filter.GetProvisionablePods(ctx, &provisioner)
+	if err != nil {
+		return reconcile.Result{Requeue: true}, err
+	}
+	if err := publishPodCounts(provisioner.Name, podsByZone, provisionablePods); err != nil {
+		// An updated value for one or more metrics was not published. Try again later.
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// 3. Schedule the next run.
+	return reconcile.Result{RequeueAfter: requeueInterval}, nil
+}
+
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	return controllerruntime.
+		NewControllerManagedBy(m).
+		Named(controllerName).
+		For(&v1alpha4.Provisioner{}, builder.WithPredicates(
+			predicate.Funcs{
+				CreateFunc:  func(_ event.CreateEvent) bool { return true },
+				DeleteFunc:  func(_ event.DeleteEvent) bool { return true },
+				UpdateFunc:  func(_ event.UpdateEvent) bool { return false },
+				GenericFunc: func(_ event.GenericEvent) bool { return false },
+			},
+		)).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: 1,
+		}).
+		Complete(c)
+}
+
+// podsByZone returns a map of slices containing all pods scheduled to nodes in each zone.
+func (c *Controller) podsByZone(ctx context.Context, provisioner *v1alpha4.Provisioner) (map[string][]v1.Pod, error) {
+	// Karpenter does not apply a label, or other marker, to pods.
+
+	results := map[string][]v1.Pod{}
+
+	// 1. Fetch all nodes associated with the provisioner -- since a label is applied
+	//    to the node.
+	nodeList := v1.NodeList{}
+	withProvisionerName := client.MatchingLabels{v1alpha4.ProvisionerNameLabelKey: provisioner.Name}
+	if err := c.KubeClient.List(ctx, &nodeList, withProvisionerName); err != nil {
+		return map[string][]v1.Pod{}, err
+	}
+
+	// 2. Get all the pods scheduled to each node and append them to the slice
+	//    associated with the node's zone.
+	for _, node := range nodeList.Items {
+		zone := node.Labels[v1.LabelTopologyZone]
+		if zone == "" {
+			return map[string][]v1.Pod{}, fmt.Errorf(
+				"Node %q provisioned by %q is missing label %s",
```

Comment:
TIL %q, cool trick. We currently have a pattern of simply naming the logger after the resource being reconciled, so you can deduplicate including the provisioner's name here.